File: 1218. Longest Arithmetic Subsequence of Given Difference/Solution.java
class Solution {
    public int longestSubsequence(int[] arr, int difference) {
        Map<Integer, Integer> dp = new HashMap<>();
        int high = 0;
        for (int a : arr) {
            dp.put(a, dp.getOrDefault(a - difference, 0) + 1);
        }
        for (int h : dp.values()) {
            high = Math.max(high, h);
        }
        return high;
    }
}

File: 1218. Longest Arithmetic Subsequence of Given Difference/Solution.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        unordered_map<int, int> dp;
        int high = 0;
        for (const auto& a : arr) {
            dp[a] = (dp.find(a - difference) == dp.end()) ? 1 : dp[a - difference] + 1;
        }
        for (const auto& [k, v] : dp) high = max(high, v);
        return high;
    }
};
